packages/functions/src/scanner.ts

```diff
@@ -0,0 +1,125 @@
+import { Config } from 'sst/node/config'
+import { Table } from 'sst/node/table'
+import { Queue } from 'sst/node/queue'
+import { Context } from 'aws-lambda'
+import { unmarshall } from '@aws-sdk/util-dynamodb'
+import { DynamoDBClient, ScanCommand, ScanCommandOutput } from '@aws-sdk/client-dynamodb'
+import { SQSClient, SendMessageCommand } from '@aws-sdk/client-sqs'
+import { LambdaClient, InvokeCommand } from '@aws-sdk/client-lambda'
+import { SSMClient, PutParameterCommand, GetParameterCommand, ParameterNotFound } from '@aws-sdk/client-ssm'
+
+const SCAN_BATCH_SIZE = 500 // max sqs msg is 256KB. each record is ~350bytes, 350 * 500 = 175KB
+const MIN_REMAINING_TIME_MS = 10_000 // threshold at which we reinvoke the lambda
+
+/**
+ * Lambda to scan an entire DynamoDB table, sending batches of records to an SQS Queue.
+ * 
+ * Stores it's progress in SSM parameter store so it can resume.
+ * 
+ * Invokes itself again when it's remaining execution time is less than `MIN_REMAINING_TIME_MS`.
+ * 
+ * Invoke it directly with 
+ *   aws lambda invoke --function-name <name> --invocation-type Event
+ * 
+ * Pass { TotalSegments: number, Segment: number} to control the table scan partition
+ * see: https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/Scan.html#Scan.ParallelScan
+ */
+export async function handler(event: any, context: Context) {
+  console.log(`invoked`, event)
+  const TotalSegments = event.TotalSegments ?? 1
+  const Segment = event.Segment ?? 0
+
+  // set via the `bind` config in the sst stack
+  const tableName = Table.srcTable.tableName
+  const queueUrl = Queue.batchQueue.queueUrl
+
+  const ssm = new SSMClient({})
+  const sqs = new SQSClient({})
+  const dynamo = new DynamoDBClient({})
+  const lambda = new LambdaClient({})
+
+  const ssmKey = `/migrate-block-index/${Config.STAGE}/last-evaluated/${TotalSegments}/${Segment}`
+  let lastEvaluated = await getLastEvaluated(ssm, ssmKey)
+
+  let recordCount = 0
+  while (true) {
+    const res: ScanCommandOutput = await dynamo.send(new ScanCommand({
+      TotalSegments,
+      Segment,
+      ExclusiveStartKey: lastEvaluated,
+      TableName: tableName,
+      Limit: SCAN_BATCH_SIZE
+    }))
+
+    if (!res.Items) {
+      console.error('Error: Scan returned no items', JSON.stringify(lastEvaluated), { TotalSegments, Segment })
+      break
+    }
+    const records = res.Items.map(i => unmarshall(i))
+    recordCount += await sendToQueue(sqs, queueUrl, records)
```

Comment:
should we skip if records is empty array?

Reply:
From my understanding from the [docs](https://docs.aws.amazon.com/amazondynamodb/latest/APIReference/API_Query.html#API_Query_ResponseSyntax), more precisely:

> If LastEvaluatedKey is not empty, it does not necessarily mean that there is more data in the result set. The only way to know when you have reached the end of the result set is when LastEvaluatedKey is empty.

it is possible that we get a page with LastEvaluatedKey that then will have no more data

Reply:
Good catch! Thanks!